### PR TITLE
Add new `--vid-stride` inference parameter for videos

### PR DESCRIPTION
**Pull request by VELCpro**
@glenn-jocher as seen in the issue https://github.com/ultralytics/yolov5/issues/8079#issue-12581169989 there's the need sometimes to skip frames while performing inference on a video.
One of the useful use cases could be batch processing of very long archive videos where high frame rates are not needed.

I added two parameters to detect.py
- --inf-rate --> number of inference given a frame rate unit
- --fr-unit --> frame rate unit ('fps', 'fpm', 'fph')

Where :
- fps = frames per second
- fpm = frames per minute
- fph = frames per hour

example use:
`python detect.py --device 0 --weights yolov5s.pt --inf-rate 1 --fr-unit fps --source vid.mp4`

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enhancement to video frame processing with the introduction of frame-rate stride control.

### 📊 Key Changes
- 🖼️ Added a `vid_stride` parameter to control the stride of video frame processing (skipping frames).
- 🛠️ `vid_stride` is implemented in `LoadImages` and `LoadStreams` classes, affecting how videos are loaded and processed.
- 🎛️ The `vid_stride` option is added as a command-line argument in both `predict.py` and `detect.py` scripts, permitting customization from the command-line interface.

### 🎯 Purpose & Impact
- 🚀 The purpose is to allow users to speed up video processing by analyzing fewer frames (skipping over frames based on stride value).
- ✨ This can significantly reduce processing time and resource usage when high frame-rate precision is not required.
- 🌐 It can also be beneficial for streaming scenarios where live feedback is more important than processing every single frame.

### CODE REVIEW

**Comment by glenn-jocher** [comment]
@VELCpro thanks for the PR! The code additions are too complicated to be added to master, especially the multiple units, but there is an example in LoadStreams with a single variable called `read` that manages frame skipping.

A couple things to remember here are that detect.py and classify/predict.py would both need to be updated, and that LoadStreams and LoadImages should share common arguments and funcitonalities, so if we exposed a new variable at the argparser it should affect both identically, i.e. `--variable 30` to read at 1 frame out of every 30, default `--variable 1`

https://github.com/ultralytics/yolov5/blob/9da6d0f9f5bc37fa386b7b82d2a963f94650949a/utils/dataloaders.py#L377-L383

**Comment by VELCpro** [comment]
@glenn-jocher thanks for the feedback!
I can try to simplify the code using only a sort of "skip_frame" variable

**Comment by glenn-jocher** [comment]
@VELCpro thanks for the updates, this looks a lot simpler now! Is this ready to merge? Have you tested on videos?

**Comment by VELCpro** [comment]
@glenn-jocher yes I tested on videos and on a folder with multiple videos and seems ok.

Now I'm on vacation with no access to my pc so i don't have the possibility to rebase the branch unfortunately. Anyway I see that the new commits in master don't refer to dataloaders.py so I think its not a problem to merge it.

**Comment by glenn-jocher** [comment]
@VELCpro looks good now. The only issue I see is that the default naming and value make it seem like we are skipping every 1 frame, i.e. defaulting to 15 FPS. Not sure if there's a better wording here like vid_stride or something. What do you think?

**Comment by VELCpro** [comment]
@glenn-jocher yes actually naming the variable skip_frame could cause some confusion.

vid_stride seems ok to me

**Comment by glenn-jocher** [comment]
@VELCpro perfect. I'll update.

**Comment by glenn-jocher** [comment]
@VELCpro PR is merged. Thank you for your contributions to YOLOv5 🚀 and Vision AI ⭐

**Comment by VELCpro** [comment]
@glenn-jocher happy to help 🙂

**Comment by Mohammad-Saad-eng** [comment]
@VELCpro @glenn-jocher Hello, hope you are doing well. I have trained a custom model on yolov8n which will detect the night traffic. I have to use it on 8 streams (or 4 in worst case) but I am facing too much lag. And when I use vid_stride, it skips the frames but the stream is still very slow. Is there a way to get better fps for that much streams using yolov8n?

**Comment by VELCpro** [comment]
Hi @Mohammad-Saad-eng everything good thank you for asking. 
What about you?

As you notice the **vid_stride** parameter allows to skip frames during the detection over a video.
This could be useful if its not needed to do detection on every single frame in real time or in doing bulk analysis on large video files.

Other ways to improve latency without touching the actual models are for example to reduce the input size of the frames given to the model for the detection.

Clearly it's a matter of trade-offs between latency and accuracy.
It's probable that you are hitting the hardware boundaries of your system with the specific yolo model selected.

What hardware you are testing the model on? 
Have you tested the model on a single stream to validate the latency?

**Comment by Mohammad-Saad-eng** [comment]
@VELCpro I am also fine.

I have tested it on Jetson Nano so far. Jetson is currently using DeepStream and TrafficCamNet for 8 streams. As number of parameters in yolov8n is significantly smaller than TrafficCamNet, I figured yolo would work fine too.

**Comment by VELCpro** [comment]
Hi @Mohammad-Saad-eng 

Me and my team are currently using a Jetson orin 32GB in conjunction with yolov8n to perform object tracking obtaining good results both in terms of accuracy and fps.

The Jetson Nano should be capable of running yolov5n (despite the limited power and the different architecture in respect to yolov8n) 

[Here](https://wiki.seeedstudio.com/YOLOv5-Object-Detection-Jetson/#pretrained-checkpoints-comparison) you can find an fps benchmark analysis performed by seeed studio  where yolov5n6 run FP16 inferences in 48ms on a Jetson Nano.

I think also they used directly the yolo docker container without deepstream.
To further enhance speed other than using **vid_stride** you can try TesorRT model format in FP16 or INT8 to speed up the computation at inference time.  [Here](https://www.seeedstudio.com/blog/2023/03/30/yolov8-performance-benchmarks-on-nvidia-jetson-devices/) a benchmark of the speed gains with different model format and jeston devices

Hope it helps

**Comment by glenn-jocher** [comment]
Hi @VELCpro-eng,

We have successfully used a Jetson Orin 32GB with yolov8n for object tracking, achieving good results in terms of both accuracy and FPS. 

Although the Jetson Nano has limited power and a different architecture compared to yolov8n, it should still be capable of running yolov5n. Seeed Studio has performed an FPS benchmark analysis, running yolov5n6 in FP16 inferences in 48ms on a Jetson Nano. You can find the analysis [here](https://wiki.seeedstudio.com/YOLOv5-Object-Detection-Jetson/#pretrained-checkpoints-comparison).

In addition to using the `vid_stride` parameter, you can also try using the TensorRT model format in FP16 or INT8 to further enhance speed during inference. Seeed Studio has conducted a benchmark testing the speed gains with different model formats on Jetson devices, which you can find [here](https://www.seeedstudio.com/blog/2023/03/30/yolov8-performance-benchmarks-on-nvidia-jetson-devices).

I hope these suggestions help.
